### PR TITLE
Update ESMF 8.3.0 to use external PIO

### PR DESCRIPTION
**Pull request by kgerheiser**
Working on https://github.com/NOAA-EMC/spack-stack/issues/201

### CODE REVIEW

File: var/spack/repos/builtin/packages/esmf/package.py [hunk 1]
```diff
@@ -263,7 +264,13 @@ def edit(self, spec, prefix):
 
             # PIO-dependent features will be enabled and will use the
             # PIO library that is included and built with ESMF.
-            os.environ['ESMF_PIO'] = 'internal'
+            if spec.satisfies('@:8.3.0'):
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/esmf/package.py]: Is that a up to 8.3.0 but not including?

**Comment by climbfuji** [comment]
First, thanks a lot for picking up this work, I was going to do it this morning but now I can do other stuff!

Good that you split it into two separate commits, this way we can feed it back cleanly to the authoritative spack repo.

The only change I'd like to make is to merge this into the release branch first. 

Oh, and let me test that on discover and run the full CI tests first.

**Comment by climbfuji** [comment]
Given the esmf 8.3.0 issues that Jun just mentioned, let's keep 8.3.0b09 and make it default again in release candidate 2.

But definitely we should get your changes in this PR in!

File: var/spack/repos/builtin/packages/esmf/package.py (hunk 1 above)
Comment (kgerheiser) [reply var/spack/repos/builtin/packages/esmf/package.py]: This might be incorrect

Comment (kgerheiser) [reply var/spack/repos/builtin/packages/esmf/package.py]: Yeah, the ranges are inclusive. I turned the logic around to if `8.3.0:`  use external PIO.

**Comment by kgerheiser** [comment]
I don't know if the logic here works with 8.3.0b09, or if the PIO change came after.

**Comment by climbfuji** [comment]
Let me check the source code ...

**Comment by mathomp4** [comment]
Just to note, we at GMAO are moving to use v8.3.0 soon (this week 🤞🏼 ?), but we use the internal PIO from ESMF. Caused some "fun" in our build system when it introduced the new `libpioc` library ESMF builds after b11 or so. Moving to using the `FindESMF.cmake` from v8.3.0 was what helped. 

I suppose, though, pointing to Ed's real PIO is probably best for features if we ever used PIO. Haven't seen a need yet...

**Comment by climbfuji** [comment]
> Just to note, we at GMAO are moving to use v8.3.0 soon (this week 🤞🏼 ?), but we use the internal PIO from ESMF. Caused some "fun" in our build system when it introduced the new `libpioc` library ESMF builds after b11 or so. Moving to using the `FindESMF.cmake` from v8.3.0 was what helped.
> 
> I suppose, though, pointing to Ed's real PIO is probably best for features if we ever used PIO. Haven't seen a need yet...

Yes, that's what we are going to do. Use one and only one version of PIO across the entire software stack.

**Comment by mathomp4** [comment]
> Yes, that's what we are going to do. Use one and only one version of PIO across the entire software stack.

Indeed. As long as ESMF builds and GEOS links to it, that's enough for me at the moment. We are mainly using the internal PIO in ESMF for the ESMF folks. If they are ever on discover, they can just point to my Baselibs and know it'll have it and it'll be whatever version is internal. 

But, as I said, we don't use it so...no harm to us! :)

**Comment by climbfuji** [comment]
> I don't know if the logic here works with 8.3.0b09, or if the PIO change came after.

I am pretty sure the PIO change came after. We'll need to handle 8.3.0b09 the "legacy way"

**Comment by kgerheiser** [comment]
Going to do a quick test, but I think this will work as is. `8.3.0:` is a higher version than `8.3.0b09` and it should not attempt to use external PIO in this case.

File: var/spack/repos/builtin/packages/esmf/package.py [hunk 2]
```diff
@@ -261,9 +263,17 @@ def edit(self, spec, prefix):
             # and NetCDF formats through ParallelIO (PIO), a third-party IO
             # software library that is integrated in the ESMF library.
 
-            # PIO-dependent features will be enabled and will use the
-            # PIO library that is included and built with ESMF.
-            os.environ['ESMF_PIO'] = 'internal'
+            # FIXME - When ESMF beta is no longer needed set this to 8.3.0:
+            if spec.satisfies('@8.3.0b12:'):
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/esmf/package.py:267]: We don't define version 8.3.0b12 in the list above

File: var/spack/repos/builtin/packages/parallelio/package.py [hunk 3]
```diff
@@ -16,8 +16,9 @@ class Parallelio(CMakePackage):
 
     maintainers = ['tkameyama']
 
-    version('2_5_4', sha256='e51dc71683da808a714deddc1a80c2650ce847110383e42f1710f3ba567e7a65')
-    version('2_5_2', sha256='935bc120ef3bf4fe09fb8bfdf788d05fb201a125d7346bf6b09e27ac3b5f345c')
+    version('2.5.7', sha256='af8af04e41af17f98f2c90b996ef0d8bcd980377e0b35e57b38938c7fdc87cbd')
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/parallelio/package.py:19]: Sigh, these PIO versions ...

File: var/spack/repos/builtin/packages/parallelio/package.py [hunk 4]
```diff
@@ -32,10 +33,15 @@ class Parallelio(CMakePackage):
...
              git='https://github.com/CESM-Development/CMake_Fortran_utils.git',
              tag='master')
+
     resource(name='genf90',
              git='https://github.com/PARALLELIO/genf90.git',
              tag='genf90_200608')
 
+    def url_for_version(self, version):
+        url = 'https://github.com/NCAR/ParallelIO/archive/refs/tags/pio{}.tar.gz'
+        return url.format(version.underscored, version)
```
Comment (climbfuji) [inline var/spack/repos/builtin/packages/parallelio/package.py]: Nice!

File: var/spack/repos/builtin/packages/esmf/package.py (hunk 2 above)
Comment (kgerheiser) [reply var/spack/repos/builtin/packages/esmf/package.py:267]: You don't need to. It's just there for comparison to `8.3.0b09` which will not match `8.3.0b12:`. Then, the line can be changed to `8.3.0:`.

Comment (climbfuji) [reply var/spack/repos/builtin/packages/esmf/package.py:267]: Oh good, so that's fine then.

**Review by climbfuji — APPROVED** [review]

**Pull request merged**